File: crabzilla-vertx-stack/src/main/java/crabzilla/stack/vertx/CommandHandlerVerticle.java
package crabzilla.stack.vertx;

import crabzilla.model.AggregateRoot;
import crabzilla.model.Command;
import crabzilla.model.Snapshot;
import crabzilla.model.UnitOfWork;
import crabzilla.model.util.Either;
import crabzilla.stack.EventRepository;
import crabzilla.stack.model.SnapshotMessage;
import io.vertx.circuitbreaker.CircuitBreaker;
import io.vertx.core.*;
import io.vertx.core.eventbus.Message;
import io.vertx.core.shareddata.LocalMap;
import lombok.NonNull;
import lombok.extern.slf4j.Slf4j;
import lombok.val;

import javax.inject.Inject;
import javax.inject.Named;
import java.util.List;
import java.util.Optional;
import java.util.function.BiFunction;
import java.util.function.Function;

import static crabzilla.model.util.Eithers.getLeft;
import static crabzilla.model.util.Eithers.getRight;
import static crabzilla.stack.vertx.util.StringHelper.commandHandlerId;
import static java.util.Collections.singletonList;

@Slf4j
public class CommandHandlerVerticle<A extends AggregateRoot> extends AbstractVerticle {

  final Class<A> aggregateRootClass;
  final Function<String, SnapshotMessage<A>> snapshotReaderFn;
  final BiFunction<Command, Snapshot<A>, Either<Exception, Optional<UnitOfWork>>> cmdHandler;
  final Function<Command, List<String>> validatorFn;
  final LocalMap<String, ShareableSnapshot<A>> cache;

  final EventRepository eventRepository;
  final Vertx vertx;
  final CircuitBreaker circuitBreaker;

  @Inject
  public CommandHandlerVerticle(@NonNull final Class<A> aggregateRootClass,
                                @NonNull final Function<String, SnapshotMessage<A>> snapshotReaderFn,
                                @NonNull final BiFunction<Command, Snapshot<A>, Either<Exception, Optional<UnitOfWork>>> cmdHandler,
                                @NonNull final Function<Command, List<String>> validatorFn,
                                @NonNull final EventRepository eventRepository,
                                @NonNull final LocalMap<String, ShareableSnapshot<A>> cache,
                                @NonNull final Vertx vertx,
                                @NonNull @Named("cmd-handler") final CircuitBreaker circuitBreaker) {
    this.aggregateRootClass = aggregateRootClass;
    this.snapshotReaderFn = snapshotReaderFn;
    this.cmdHandler = cmdHandler;
    this.validatorFn = validatorFn;
    this.eventRepository = eventRepository;
    this.cache = cache;
    this.vertx = vertx;
    this.circuitBreaker = circuitBreaker;
  }

  @Override
  public void start() throws Exception {

    vertx.eventBus().consumer(commandHandlerId(aggregateRootClass), msgHandler());

  }

  Handler<Message<Command>> msgHandler() {

    return (Message<Command> msg) -> {

      vertx.executeBlocking((Future<CommandExecution> future) -> {

        val command = msg.body();

        if (command==null) {
          future.complete(CommandExecution.VALIDATION_ERROR(singletonList("Command cannot be null. Check if JSON payload is valid.")));
          return;
        }

        log.info("received a command {}", command);

        val constraints = validatorFn.apply(command);

        if (!constraints.isEmpty()) {
          future.complete(CommandExecution.VALIDATION_ERROR(command.getCommandId(), constraints));
          return;
        }

        circuitBreaker.fallback(throwable -> {

          log.error("Fallback for command " + command.getCommandId(), throwable);
          return CommandExecution.FALLBACK(command.getCommandId()); })

        .execute(cmdHandler(command))

        .setHandler(resultHandler(msg));

      }, false, resultHandler(msg));

    };
  }

  Handler<Future<CommandExecution>> cmdHandler(final Command command) {

    return future -> {

      // TODO this call should be a Future ?
      val snapshotDataMsg = snapshotReaderFn.apply(command.getTargetId().getStringValue());

      if (snapshotDataMsg.hasNewSnapshot()) {
        cache.put(command.getCommandId().toString(), new ShareableSnapshot<>(snapshotDataMsg.getSnapshot()));
      }

      final Either<Exception, Optional<UnitOfWork>> either = cmdHandler.apply(command, snapshotDataMsg.getSnapshot());

      val optException = getLeft(either);

      if (optException.isPresent()) {
        log.error("Business logic error for command " + command.getCommandId(), optException.get());
        future.complete(CommandExecution.BUSINESS_ERROR(command.getCommandId()));
        return;
      }

      val optUnitOfWork = getRight(either);

      if (!optUnitOfWork.get().isPresent()) {
        future.complete(CommandExecution.UNKNOWN_COMMAND(command.getCommandId()));
        return;
      }

      val uow = optUnitOfWork.get().get();
      val uowSequence = eventRepository.append(uow);
      val cmdHandleResp = CommandExecution.SUCCESS(uow, uowSequence);

      future.complete(cmdHandleResp);

    };

  }

  Handler<AsyncResult<CommandExecution>> resultHandler(Message<Command> msg) {

    return (AsyncResult<CommandExecution> resultHandler) -> {

      if (resultHandler.succeeded()) {

        val resp = resultHandler.result();
        log.info("success: {}", resp);
        msg.reply(resp);

      } else {

        log.error("error cause: {}", resultHandler.cause());
        log.error("error message: {}", resultHandler.cause().getMessage());
        resultHandler.cause().printStackTrace();
        // TODO customize conform commandResult
        msg.fail(400, resultHandler.cause().getMessage());
      }

    };
  }
}

File: crabzilla-core/src/main/java/io/github/crabzilla/CommandHandlerVerticle.kt
package io.github.crabzilla

import io.github.crabzilla.CommandExecution.RESULT
import io.vertx.core.AbstractVerticle
import io.vertx.core.Future
import io.vertx.core.Handler
import io.vertx.core.eventbus.DeliveryOptions
import org.slf4j.LoggerFactory

class CommandHandlerVerticle<A : Entity>(val name: String,
                                         private val seedValue: A,
                                         private val cmdHandler: (Command, Snapshot<A>) -> CommandResult?,
                                         private val validatorFn: (Command) -> List<String>,
                                         private val eventJournal: UnitOfWorkRepository,
                                         private val snapshotRepo: SnapshotRepository<A>)
  : AbstractVerticle() {

  companion object {
    internal val log = LoggerFactory.getLogger(CommandHandlerVerticle::class.java)
    internal val options = DeliveryOptions().setCodecName("CommandExecution")
  }

  @Throws(Exception::class)
  override fun start() {

    log.info("starting command verticle for $name")

    vertx.eventBus().consumer<Command>(cmdHandlerEndpoint(name), Handler { commandMsg ->

      val command = commandMsg.body()

      log.info("received a command $command")
      val constraints = validatorFn.invoke(command)

      if (!constraints.isEmpty()) {
        val result = CommandExecution(commandId = command.commandId, result = RESULT.VALIDATION_ERROR,
          constraints = constraints)
        commandMsg.reply(result)
        return@Handler
      }

      val targetId = command.targetId.value()

      snapshotRepo.retrieve(targetId, name, Handler { fromCacheResult ->

        if (fromCacheResult.failed()) {
          val result = CommandExecution(commandId = command.commandId, result = RESULT.HANDLING_ERROR)
          commandMsg.reply(result)
          return@Handler
        }

        val snapshotFromCache = if (fromCacheResult.failed()) null else fromCacheResult.result()
        val cachedSnapshot = snapshotFromCache ?: Snapshot(seedValue, 0)

        val commandResult: CommandResult? = cmdHandler.invoke(command, cachedSnapshot)

        if (commandResult?.unitOfWork == null) {
          val result = CommandExecution(commandId = command.commandId, result = RESULT.HANDLING_ERROR)
          commandMsg.reply(result)
          return@Handler
        }

        commandResult.inCaseOfSuccess { uow ->

          val appendFuture = Future.future<Int>()

          eventJournal.append(uow!!, appendFuture, name)

          appendFuture.setHandler { appendAsyncResult ->

            if (appendAsyncResult.failed()) {
              val error = appendAsyncResult.cause()
              log.error("appendUnitOfWork for command " + command.commandId, error.message)
              val result = if (error is DbConcurrencyException) {
                CommandExecution(commandId = command.commandId, result = RESULT.CONCURRENCY_ERROR)
              } else {
                CommandExecution(commandId = command.commandId, result = RESULT.HANDLING_ERROR)
              }
              commandMsg.reply(result)
              return@setHandler
            }

            val uowSequence = appendAsyncResult.result()
            log.info("uowSequence: {}", uowSequence)
            commandMsg.reply(CommandExecution(result = RESULT.SUCCESS, commandId = command.commandId,
              unitOfWork = uow, uowSequence = uowSequence), options)
          }

        }

        commandResult.inCaseOfError { error ->
          log.info("error: {}", error)
          commandMsg.reply(CommandExecution(commandId = command.commandId, result = RESULT.HANDLING_ERROR))
          return@inCaseOfError
        }

      })

    })

  }

  class DbConcurrencyException(s: String) : RuntimeException(s)

}
